flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServices.java

```diff
@@ -311,7 +314,12 @@ public static TaskManagerServices fromConfiguration(
 			permanentBlobService,
 			BlobLibraryCacheManager.defaultClassLoaderFactory(
 				taskManagerServicesConfiguration.getClassLoaderResolveOrder(),
-				taskManagerServicesConfiguration.getAlwaysParentFirstLoaderPatterns()));
+				taskManagerServicesConfiguration.getAlwaysParentFirstLoaderPatterns(),
+				classLoadingException -> {
+					if (ExceptionUtils.isMetaspaceOutOfMemoryError(classLoadingException)) {
+						fatalErrorHandler.onFatalError(classLoadingException);
+					}
```

Comment:
I thought we agreed that we wanted to try a graceful TM shutdown with reporting to JM before the JVM fatal exit. This will be also not flexible enough to behave differently in JM (related to the previous comment). If we want to have only one way of handling then indeed it can be simplified.

Reply:
We can introduce an advanced user class loader option to decide whether to fail on Metaspace OOM on user class loading to fallback to the existing behaviour if needed. The class loading will fail on OOM, by default.